modules/engine/src/main/scala/com/gsk/kg/engine/Multiset.scala

```diff
@@ -37,40 +42,85 @@ final case class Multiset(
     * @param other
     * @return the join result of both multisets
     */
-  def join(other: Multiset): Multiset =
-    (this, other) match {
-      case (a, b) if a.isEmpty => b
-      case (a, b) if b.isEmpty => a
-      case (Multiset(aBindings, aDF), Multiset(bBindings, bDF))
-          if aBindings.intersect(bBindings).isEmpty =>
-        val df = aDF
-          .as("a")
-          .crossJoin(bDF.as("b"))
-        Multiset(aBindings.union(bBindings), df)
-      case (a, b) =>
-        val common = a.bindings.intersect(b.bindings)
-        val df = a.dataframe
-          .as("a")
-          .join(b.dataframe.as("b"), common.map(_.s).toSeq)
-        Multiset(a.bindings.union(b.bindings), df)
-    }
+  def join(other: Multiset): Multiset = (this, other) match {
+    case (l, r) if l.isEmpty => r
+    case (l, r) if r.isEmpty => l
+    case (l, r)
+        if (filterGraph(l).bindings intersect filterGraph(
+          r
+        ).bindings).isEmpty =>
```

Comment:
Can we extract the guard to a named private function? something like
```suggestion
        if noCommonBindingsBetween(l, r) =>
```

Reply:
Yeah, good idea